sql-plugin/src/main/scala/org/apache/spark/sql/rapids/predicates.scala

```diff
@@ -70,63 +65,52 @@ case class GpuAnd(left: Expression, right: Expression) extends CudfBinaryOperato
 
   override def binaryOp: BinaryOp = BinaryOp.LOGICAL_AND
 
-  import GpuLogicHelper._
-
-  // The CUDF implementation of `and` will return a null if either input is null
-  // Spark does not.
-  // |LHS/RHS| TRUE  | FALSE | NULL  |
-  // |TRUE   | TRUE  | FALSE | NULL  |
-  // |FALSE  | FALSE | FALSE | FALSE |
-  // |NULL   | NULL  | FALSE | NULL  |
-  // So we have to make some adjustments.
-  //  IF (A <=> FALSE, FALSE,
-  //    IF (B <=> FALSE, FALSE,
-  //      A cudf_and B))
-
   override def doColumnar(lhs: GpuColumnVector, rhs: GpuColumnVector): ColumnVector = {
     val l = lhs.getBase
     val r = rhs.getBase
-    withResource(Scalar.fromBool(false)) { falseVal =>
-      val firstPass = withResource(l.and(r)) { lAndR =>
-        withResource(l.equalToNullAware(falseVal)) { lIsFalse =>
-          lIsFalse.ifElse(falseVal, lAndR)
-        }
-      }
-      withResource(firstPass) { firstPass =>
-        withResource(r.equalToNullAware(falseVal)) { rIsFalse =>
-          rIsFalse.ifElse(falseVal, firstPass)
+    withResource(new Table(l, r)) {
+      table => {
+        val astOp = new ast.BinaryOperation(ast.BinaryOperator.NULL_LOGICAL_AND,
+          new ast.ColumnReference(0),
+          new ast.ColumnReference(1))
+        withResource(astOp.compile()) { compiledAstExpr =>
+          compiledAstExpr.computeColumn(table)
         }
       }
     }
   }
 
   override def doColumnar(lhs: GpuScalar, rhs: GpuColumnVector): ColumnVector = {
-    val l = lhs.getBase
     val r = rhs.getBase
-    withResource(Scalar.fromBool(false)) { falseVal =>
-      if (eqNullAware(lhs, false)) {
-        ColumnVector.fromScalar(falseVal, r.getRowCount.toInt)
-      } else {
-        withResource(l.and(r)) { lAndR =>
-          withResource(r.equalToNullAware(falseVal)) { rIsFalse =>
-            rIsFalse.ifElse(falseVal, lAndR)
-          }
+    val lhsLit = if (lhs.isValid) {
```

Comment:
Getting an AST Literal from a `GpuScalar` should be a utility method, as it appears multiple times here and we may want to update all occurrences in the future if Java cudf ASTs are updated to take `Scalar` instances directly.

Reply:
Could GpuLiteral be used here?

Reply:
GpuLiteral is something that can appear in the query plans and represents a query literal that we could translate to the GPU.  However in this case we're handed a GpuScalar as a method parameter, not a GpuLiteral.  If you're asking simply where would the utility method be placed, I think GpuScalar is more appropriate than GpuLiteral since the former is a much closer match to a libcudf AST literal (which is just a wrapper for a GPU scalar value like GpuScalar is).